.cbmc-batch/include/proof_helpers/make_common_data_structures.h

```diff
@@ -133,3 +133,19 @@ void ensure_allocated_hash_table(struct aws_hash_table *map, size_t max_table_en
  * Makes a valid c string, with as much nondet as possible, len < max
  */
 const char *make_arbitrary_c_str(size_t max_size);
+
+/**
+ * A correct hash table has max_load < size.  This means that there is always one slot empty.
+ * This function is useful for assuming that there is some (nondet) slot which is empty
+ * which is necessary to prove termination for hash-table deletion code.  Should only be used inside
+ * an assume because of the way it does nondet.
+ */
+bool aws_hash_table_has_an_empty_slot(struct aws_hash_table *map, size_t *rval);
```

Comment:
* Suggest `struct aws_hash_table const * const map, size_t * const rval`
* Naiive question: are you checking elsewhere that `max_load < size`? If that condition doesn't hold, what actually happens–does the proof break, or might it silently let bugs through?

Reply:
1. done
1. That is checked as a pre-condition to the function, which translates to an ```assert``` at runtime.  I'm also checking that every function which acts on hash-tables does that as a post-condition, so once I finish those proofs we will have a functional proof of this property.